### ISSUE TITLE
Data source template_file does not respect changed resources in variables

### ISSUE THREAD

**Issue by littldr** [hashicorp/terraform#18724]
### Terraform Version
```
v0.11.7
```

### Terraform Configuration Files
```hcl
provider "aws" {}

data "template_file" "something" {
  template = "$${cluster_name}"

  vars {
    cluster_name = "${aws_ecs_cluster.cluster.name}"
  }
}

resource "aws_ecs_cluster" "cluster" {
  name = "foobar"
}

output "template_file" {
  value = "${data.template_file.something.rendered}"
}

```

### Expected Behavior
Execute following steps:
1. `terraform init`
1. `terraform apply`
2. Change `aws_ecs_cluster.cluster.name` to something different (e.g. `foobar_superbar`)
3. `terraform apply`

Expected to see:
1. Apply
```
Outputs:

template_file = foobar
```
2. Apply
```
Outputs:

template_file = foobar_superbar
```

### Actual Behavior
The `template_file` output does not change as expected.

1. Apply
```
Outputs:

template_file = foobar
```
2. Apply
```
Outputs:

template_file = foobar
```

### Steps to Reproduce
Please list the full steps required to reproduce the issue, for example:

1. `terraform init`
2. `terraform apply`
3. Change `aws_ecs_cluster.cluster.name` to something different (e.g. `foobar_superbar`)
4. `terraform apply`

### Additional Context
When using the deprecated resource `template_file` everything works as expected.
```hcl
resource "template_file" "something" {
  template = "$${cluster_name}"

  vars {
    cluster_name = "${aws_ecs_cluster.cluster.name}"
  }
}
```

**Comment by apparentlymart** [comment]
Hi @littldr! Sorry for the slow response.

This is working as currently designed, but I agree that the current design is confusing. The root problem here is that Terraform currently considers "refresh" and "plan" to be two separate steps (both run in sequence inside the `terraform apply` command), with data resources being processed during the refresh step but the changes to managed resources (`resource` blocks) in configuration only processed during the planning step.

This design causes some other similar issues, such as when a provider configuration depends on a managed resource configuration but the refresh phase will use its prior value: this is sometimes the right answer, but often not.

In a future release we intend to combine these two phases together so that refreshing and planning can be interleaved, getting the behavior you were looking for here. This will require some refactoring of moderate complexity and so we need to complete some more design and prototyping first to figure out the details. Since our current focus is on the configuration language improvements for Terraform 0.12 we won't be able to do any detailed investigation on this for the moment but we'll return to this in future.

Thanks for pointing this out! Although the general problem here was known to us, this issue captures well a specific, straightforward use-case which we can use to represent the general problem.

**Comment by damdr** [comment]
hello , same issue with v0.11.10

**Comment by JamesThorpe** [comment]
Is this related to [an issue I'm seeing](https://stackoverflow.com/questions/53540407/updating-a-count-in-a-template-file-is-not-affecting-the-output) also?

**Comment by tullydwyer** [comment]
Hi all, having the same issue here. Is there any possible way to force a refresh of the data source? I cant seem to find a single way to force a refresh with updated values.

Thanks

**Comment by brentarias** [comment]
I had this problem when I had my template_file supply a var which constituted a circular-reference.  That is, the template_file was feeding a policy associated with an AWS Lambda, but at the same time it was trying to supply the Lambda's ARN as a parameter to the template_file.  TF did not tell me there was a circular reference.  Instead template_file simply rendered an empty policy for the Lambda, and the Lambda could not be created.  
I mistakenly thought I had made a mistake in the policy file which the template_file was pointed at.  So I kept making changes in the policy file and none of those change impacted anything.  It was only after I removed the circular dependency that "normalcy" returned.
